Smoke test Runtime Exception: "Already contains key checkTModuleCreated with child org.eclipse.n4js.smith.TimedDataCollector"

This exception appeared when running the smoke tests locally (GeneratedSmokeTestCases3, test_0602 and test_0603). I never saw it on Jenkins.

**stacktrace**
```
java.lang.AssertionError: java.lang.RuntimeException: Already contains key checkTModuleCreated with child org.eclipse.n4js.smith.TimedDataCollector@78ac9fa3
	at org.eclipse.n4js.ExceptionAwareDiagnosticConverter.handleExceptions(ExceptionAwareDiagnosticConverter.java:49)
	at org.eclipse.n4js.ExceptionAwareDiagnosticConverter.convertValidatorDiagnostic(ExceptionAwareDiagnosticConverter.java:38)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.issueFromEValidatorDiagnostic(ResourceValidatorImpl.java:180)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:149)
	at org.eclipse.n4js.validation.N4JSResourceValidator.validate(N4JSResourceValidator.java:97)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at org.eclipse.n4js.validation.N4JSResourceValidator.validate(N4JSResourceValidator.java:77)
	at org.eclipse.n4js.analysis.ExceptionAnalyser.getScriptErrors(ExceptionAnalyser.java:77)
	at org.eclipse.n4js.analysis.PositiveAnalyser.analyse(PositiveAnalyser.java:57)
	at org.eclipse.n4js.smoke.tests.Generated3SmokeTest$DefaultSmokeTester.doProcessFile(Generated3SmokeTest.java:47)
	at org.eclipse.n4js.smoke.tests.UniqueScenarioProcessor.processFile(UniqueScenarioProcessor.java:50)
	at org.eclipse.xtext.testing.smoketest.DeltaScenarioProcessor$Adapter.processFile(DeltaScenarioProcessor.java:52)
	at org.eclipse.xtext.testing.smoketest.DeltaScenarioProcessor.processFile(DeltaScenarioProcessor.java:33)
	at org.eclipse.xtext.testing.smoketest.Scenario$1.processInput(Scenario.java:40)
	at org.eclipse.xtext.testing.smoketest.Scenario.processInput(Scenario.java:155)
	at org.eclipse.xtext.testing.smoketest.internal.ParallelPermutingScenarioRunner.doProcess(ParallelPermutingScenarioRunner.java:42)
	at org.eclipse.xtext.testing.smoketest.internal.AbstractParallelScenarioRunner.process(AbstractParallelScenarioRunner.java:168)
	at org.eclipse.xtext.testing.smoketest.internal.AbstractParallelScenarioRunner$4.evaluate(AbstractParallelScenarioRunner.java:155)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.eclipse.xtext.testing.smoketest.internal.AbstractParallelScenarioRunner.runChild(AbstractParallelScenarioRunner.java:93)
	at org.eclipse.xtext.testing.smoketest.internal.AbstractParallelScenarioRunner$2.run(AbstractParallelScenarioRunner.java:121)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: Already contains key checkTModuleCreated with child org.eclipse.n4js.smith.TimedDataCollector@78ac9fa3
	at org.eclipse.n4js.smith.TimedDataCollector.addChild(TimedDataCollector.java:91)
	at org.eclipse.n4js.smith.DataCollectors.get(DataCollectors.java:109)
	at org.eclipse.n4js.smith.DataCollectors.getOrCreateDataCollector(DataCollectors.java:78)
	at org.eclipse.n4js.validation.N4JSValidator$N4JSMethodWrapperCancelable.invoke(N4JSValidator.java:105)
	at org.eclipse.xtext.validation.AbstractDeclarativeValidator.internalValidate(AbstractDeclarativeValidator.java:312)
	at org.eclipse.xtext.validation.AbstractInjectableValidator.validate(AbstractInjectableValidator.java:71)
	at org.eclipse.xtext.validation.CompositeEValidator.validate(CompositeEValidator.java:151)
	at org.eclipse.emf.ecore.util.Diagnostician.doValidate(Diagnostician.java:171)
	at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:158)
	at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:137)
	at org.eclipse.xtext.validation.CancelableDiagnostician.validate(CancelableDiagnostician.java:39)
	at org.eclipse.n4js.utils.validation.PrePostDiagnostician.validate(PrePostDiagnostician.java:56)
	at org.eclipse.emf.ecore.util.Diagnostician.validate(Diagnostician.java:120)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:146)
	... 22 more

```

--- a/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
+++ b/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
@@ -31,8 +31,6 @@ public enum DataCollectors {
 
 	private final Map<String, DataCollector> collectors = new HashMap<>();
 	private final AtomicBoolean pauseAllCollectors = new AtomicBoolean(true);
-	/** Deal with multiple threads competing to access {@link DataCollectors#collectors} concurrently. */
-	private final Object lock = new Object();
 
 	/**
 	 * returns existing collector for the provided key. If there is no collector corresponding to the key creates new
@@ -58,28 +56,7 @@ public enum DataCollectors {
 			throw new RuntimeException("DataCollector key cannot be null or empty");
 		}
 
-		DataCollector parent = null;
-		// acquire lock to avoid other thread modifying {@link collectors}
-		synchronized (lock) {
-			if (parentKeys != null) {
-				String parentKey = parentKeys[0];
-				String prevParentKey = parentKey;
-				parent = collectors.get(parentKey);
-				if (parent == null)
-					throw new RuntimeException("Cannot find parent for key " + parentKey);
-				// we have root parent, iterate over its children
-				for (int i = 1; i < parentKeys.length; i++) {
-					parentKey = parentKeys[i];
-					parent = parent.getChild(parentKey);
-					// since we have parent keys we break at key with no mapping to parent
-					if (parent == null)
-						throw new RuntimeException(
-								"Cannot find collector for key " + parentKey + " and parent " + prevParentKey);
-					prevParentKey = parentKey;
-				}
-			}
-		}
-
+		DataCollector parent = getParent(parentKeys);
 		return get(key, parent);
 	}
 
@@ -96,26 +73,47 @@ public enum DataCollectors {
 		return get(key, parent);
 	}
 
-	private synchronized DataCollector get(String key, DataCollector parent) {
-		DataCollector collector = null;
-		// acquire lock to avoid other thread modifying {@link collectors}
-		synchronized (lock) {
-			if (parent == null) {
-				collector = collectors.get(key);
-				if (collector == null) {
-					collector = new TimedDataCollector();
-					collector.setPaused(this.pauseAllCollectors.get());
-					collectors.put(key, collector);
-				}
-			} else {
-				collector = parent.getChild(key);
-				if (collector == null) {
-					collector = new TimedDataCollector();
-					collector.setPaused(this.pauseAllCollectors.get());
-					parent.addChild(key, collector);
-				}
+	private synchronized DataCollector getParent(String... parentKeys) {
+		DataCollector parent = null;
+		if (parentKeys != null) {
+			String parentKey = parentKeys[0];
+			String prevParentKey = parentKey;
+			parent = collectors.get(parentKey);
+			if (parent == null)
+				throw new RuntimeException("Cannot find parent for key " + parentKey);
+			// we have root parent, iterate over its children
+			for (int i = 1; i < parentKeys.length; i++) {
+				parentKey = parentKeys[i];
+				parent = parent.getChild(parentKey);
+				// since we have parent keys we break at key with no mapping to parent
+				if (parent == null)
+					throw new RuntimeException(
+							"Cannot find collector for key " + parentKey + " and parent " + prevParentKey);
+				prevParentKey = parentKey;
 			}
 		}
+		return parent;
+	}
+
+	private synchronized DataCollector get(String key, DataCollector parent) {
+		DataCollector collector = null;
+
+		if (parent == null) {
+			collector = collectors.get(key);
+			if (collector == null) {
+				collector = new TimedDataCollector();
+				collector.setPaused(this.pauseAllCollectors.get());
+				collectors.put(key, collector);
+			}
+		} else {
+			collector = parent.getChild(key);
+			if (collector == null) {
+				collector = new TimedDataCollector();
+				collector.setPaused(this.pauseAllCollectors.get());
+				parent.addChild(key, collector);
+			}
+		}
+
 		return collector;
 	}
 

--- a/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
+++ b/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
@@ -14,6 +14,8 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
 
 import com.google.common.base.Strings;
 
@@ -31,6 +33,8 @@ public enum DataCollectors {
 
 	private final Map<String, DataCollector> collectors = new HashMap<>();
 	private final AtomicBoolean pauseAllCollectors = new AtomicBoolean(true);
+	/** Deal with multiple threads competing to access {@link DataCollectors#collectors} concurrently. */
+	private final Lock lock = new ReentrantLock(true);
 
 	/**
 	 * returns existing collector for the provided key. If there is no collector corresponding to the key creates new
@@ -57,22 +61,28 @@ public enum DataCollectors {
 		}
 
 		DataCollector parent = null;
-		if (parentKeys != null) {
-			String parentKey = parentKeys[0];
-			String prevParentKey = parentKey;
-			parent = collectors.get(parentKey);
-			if (parent == null)
-				throw new RuntimeException("Cannot find parent for key " + parentKey);
-			// we have root parent, iterate over its children
-			for (int i = 1; i < parentKeys.length; i++) {
-				parentKey = parentKeys[i];
-				parent = parent.getChild(parentKey);
-				// since we have parent keys we break at key with no mapping to parent
+		// acquire lock to avoid other thread modifying {@link collectors}
+		try {
+			lock.lock();
+			if (parentKeys != null) {
+				String parentKey = parentKeys[0];
+				String prevParentKey = parentKey;
+				parent = collectors.get(parentKey);
 				if (parent == null)
-					throw new RuntimeException(
-							"Cannot find collector for key " + parentKey + " and parent " + prevParentKey);
-				prevParentKey = parentKey;
+					throw new RuntimeException("Cannot find parent for key " + parentKey);
+				// we have root parent, iterate over its children
+				for (int i = 1; i < parentKeys.length; i++) {
+					parentKey = parentKeys[i];
+					parent = parent.getChild(parentKey);
+					// since we have parent keys we break at key with no mapping to parent
+					if (parent == null)
+						throw new RuntimeException(
+								"Cannot find collector for key " + parentKey + " and parent " + prevParentKey);
+					prevParentKey = parentKey;
+				}
 			}
+		} finally {
+			lock.unlock();
 		}
 
 		return get(key, parent);
@@ -91,23 +101,28 @@ public enum DataCollectors {
 		return get(key, parent);
 	}
 
-	private DataCollector get(String key, DataCollector parent) {
-
+	private synchronized DataCollector get(String key, DataCollector parent) {
 		DataCollector collector = null;
-		if (parent == null) {
-			collector = collectors.get(key);
-			if (collector == null) {
-				collector = new TimedDataCollector();
-				collector.setPaused(this.pauseAllCollectors.get());
-				collectors.put(key, collector);
+		// acquire lock to avoid other thread modifying {@link collectors}
+		try {
+			this.lock.lock();
+			if (parent == null) {
+				collector = collectors.get(key);
+				if (collector == null) {
+					collector = new TimedDataCollector();
+					collector.setPaused(this.pauseAllCollectors.get());
+					collectors.put(key, collector);
+				}
+			} else {
+				collector = parent.getChild(key);
+				if (collector == null) {
+					collector = new TimedDataCollector();
+					collector.setPaused(this.pauseAllCollectors.get());
+					parent.addChild(key, collector);
+				}
 			}
-		} else {
-			collector = parent.getChild(key);
-			if (collector == null) {
-				collector = new TimedDataCollector();
-				collector.setPaused(this.pauseAllCollectors.get());
-				parent.addChild(key, collector);
-			}
+		} finally {
+			lock.unlock();
 		}
 		return collector;
 	}

--- a/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
+++ b/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
@@ -14,8 +14,6 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
 
 import com.google.common.base.Strings;
 
@@ -34,7 +32,7 @@ public enum DataCollectors {
 	private final Map<String, DataCollector> collectors = new HashMap<>();
 	private final AtomicBoolean pauseAllCollectors = new AtomicBoolean(true);
 	/** Deal with multiple threads competing to access {@link DataCollectors#collectors} concurrently. */
-	private final Lock lock = new ReentrantLock(true);
+	private final Object lock = new Object();
 
 	/**
 	 * returns existing collector for the provided key. If there is no collector corresponding to the key creates new
@@ -62,8 +60,7 @@ public enum DataCollectors {
 
 		DataCollector parent = null;
 		// acquire lock to avoid other thread modifying {@link collectors}
-		try {
-			lock.lock();
+		synchronized (lock) {
 			if (parentKeys != null) {
 				String parentKey = parentKeys[0];
 				String prevParentKey = parentKey;
@@ -81,8 +78,6 @@ public enum DataCollectors {
 					prevParentKey = parentKey;
 				}
 			}
-		} finally {
-			lock.unlock();
 		}
 
 		return get(key, parent);
@@ -104,8 +99,7 @@ public enum DataCollectors {
 	private synchronized DataCollector get(String key, DataCollector parent) {
 		DataCollector collector = null;
 		// acquire lock to avoid other thread modifying {@link collectors}
-		try {
-			this.lock.lock();
+		synchronized (lock) {
 			if (parent == null) {
 				collector = collectors.get(key);
 				if (collector == null) {
@@ -121,8 +115,6 @@ public enum DataCollectors {
 					parent.addChild(key, collector);
 				}
 			}
-		} finally {
-			lock.unlock();
 		}
 		return collector;
 	}

--- a/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
+++ b/plugins/org.eclipse.n4js.smith/src/org/eclipse/n4js/smith/DataCollectors.java
@@ -124,7 +124,7 @@ public enum DataCollectors {
 	}
 
 	/** sets {@code paused} state for all data collectors. */
-	void setPaused(boolean paused) {
+	synchronized void setPaused(boolean paused) {
 		this.pauseAllCollectors.set(paused);
 		collectors.values().forEach(collector -> collector.setPaused(paused));
 	}

